substrate/runtime/contract/src/lib.rs

```diff
@@ -142,20 +153,98 @@ impl<'a, T: Ext + 'a> Runtime<'a, T> {
 			}
 		}
 	}
+
+	/// Save a data buffer as a result of the execution.
+	///
+	/// This function also charges gas for the returning.
+	///
+	/// Returns `Err` if there is not enough gas.
+	fn store_return_data(&mut self, data: Vec<u8>) -> Result<(), ()> {
+		let price = (self.config.return_data_per_byte_cost as u64)
+			.checked_mul(data.len() as u64)
+			.ok_or_else(|| ())?;
+		if self.charge_gas(price) {
+			self.special_trap = Some(SpecialTrap::Return(data));
+			Ok(())
+		} else {
+			Err(())
+		}
+	}
+}
+
+fn to_execution_result<T: Ext>(
+	runtime: Runtime<T>,
+	run_err: Option<sandbox::Error>,
+) -> Result<ExecutionResult, Error> {
+	let gas_left = runtime.gas_limit - runtime.gas_used;
```

Comment:
`overflowing_sub` or make comment that it is guaranteed to not overflow? 
since it is unclear now

Reply:
that's a good idea!

Reply:
Done https://github.com/paritytech/polkadot/pull/253/commits/aa5c02a663de3511794f8101eeacb5b5414883e4 !